torch/nn/parallel/distributed.py

```diff
@@ -697,7 +701,12 @@ def forward(self, *inputs, **kwargs):
             # this forward pass, to ensure we short circuit reduction for any
             # unused parameters. Only if `find_unused_parameters` is set.
             if self.find_unused_parameters:
-                self.reducer.prepare_for_backward(list(_find_tensors(output)))
+                if RPC_AVAILABLE and isinstance(self.module, Pipe):
+                    # Unwrap RRef to get real output for Pipe.
+                    # TODO: Needs to be reworked for cross host pipelining.
+                    self.reducer.prepare_for_backward(list(_find_tensors(output.local_value())))
```

Comment:
Is the RRef always guaranteed to be owned by the node that this runs on? How exactly do we guarantee this?

Reply:
Also, is this the only special case that we're going to have where we have to unwrap to get the actual tensors needed in `prepare_for_backwards`? I'm a bit concerned about too tightly coupling Pipe with DDP, when they should probably be as independent as they can be (i.e., ideally DDP won't have to deal with pipe-specific logic). Could we eventually refactor to make that so?

Reply:
> Is the RRef always guaranteed to be owned by the node that this runs on? How exactly do we guarantee this?

This kinda works right now since Pipe is single host only, but as I mentioned in the comment above, this needs to be reworked for cross host pipelining.

>  I'm a bit concerned about too tightly coupling Pipe with DDP, when they should probably be as independent as they can be (i.e., ideally DDP won't have to deal with pipe-specific logic). Could we eventually refactor to make that so?

This is a good point and when I thought about it a bit more, I think we can make this independent of Pipe for now and also more generic. What we can do is enhance `find_tensors` to unwrap a local RRef and look for tensors inside if it encounters a local RRef.